<center> <H1>第六章 函式與函式庫</H1> </center>

## 6.1 Python函式概論


函式為每一個程式語言共同的部份，可以視為程式開發功能的基本元件。一般而言，可分為內建函式(Built-in functions)及使用者自訂函式(user-defined functions)。內建函式可以是標準或第三方程式庫的組成，自訂函式則可用來解決使用這所遭遇的特定問題。

程式中建立及運用函式，可以想像成數學的函數，`y = f(x)`。f(x)是一個函數，如果給予適當的參數(x)，f(x)可以計算產生一個結果。程式中的函式也是如此，函式可視為一個黑盒子，給予它「輸入」(input)，經過黑盒子運作，產生可用的「輸出」(output)。

函式是程式語言的一個結構化元件，它聚集一些程式碼，用來執行某一特定程式工作。函式使我們能夠重複使用程式碼，而不需要一再地拷貝及貼上程式。運用函式得宜，可以使撰寫的程式容易理解，使程式具有較佳的品質，且降低開發與維護程式成本。在不同的程式語言中，函式有不同的名稱，如副程式、方法、程序、...等。

建立函式需要`def`的關鍵字，定義函式的語法如下

```python
def 函式名稱(參數列):
    程式敘述
```
程式敘述可稱為函式主體(function body)，注意必須遵守內縮規則。參數列可有可無，即可以無參數或多個參數，如果有參數列定義，必須遵循參數資料型態及個數，否則會產生韓式呼叫錯誤。函式可以回傳結果值，回傳結果須使用`return`關鍵字。當函式主體碰到`return`關鍵字時，函式執行停止，回傳函式執行結果，執行結果在`return`之後，可以是常數、變數、運算式...等。

### 函式典型呼叫

In [2]:
def hello(friend):
    print("Hello", friend)
    
hello("Jason")
hello("Mary")
hello("Daniel")

Hello Jason
Hello Mary
Hello Daniel


上例中，hello(friend)是一個函式，hello為函式名稱，friend為輸入參數。`print("Hello", friend)`是函式主體，必須符合內縮規則，第1, 2行程式為函式定義，4, 5, 6行程式為函式呼叫(invoke)。例如，第4行呼叫函式，傳入參數「Jason」，程式進入函式第2行執行函式主體，列印「Hello Jason」。

### 區域變數

在函式內部出現的變數（包含輸入參數）稱為區域變數(Local variable)。區域變數的有效範圍只有在函式內部，離開函式就無法作用。如上例中，若在函式外部存取friend變數將產生「變數名稱未定義」(name 'friend' is not defined)錯誤。

In [3]:
def hello(friend):
    print("Hello", friend)
    
hello("Jason")
hello("Mary")
hello("Daniel")
print(friend)

Hello Jason
Hello Mary
Hello Daniel


NameError: name 'friend' is not defined

## 6.2 內建函式

隨著Pyhton程式語言安裝，Python提供一些有用的函式，這些函式稱為內建函式(built-in functions)。例如，常用的列印執行結果的`print()`函式、可以轉換整數資料型別的`int()`函式等。Pyhton程式語言3.6版本提供68個內建函式，請參考[內建函式一欄表](https://www.programiz.com/python-programming/methods/built-in)。使用內建函式時，應注意函式的輸入參數定義及輸出定義與格式。

<center>**轉換函式**</center>

名稱   |  功能  |  範例  |  結果
:----|:-------|--------|-------|
int()   | 整數轉換 | int('99')  |  99
float() | 浮點數轉換   |  float('3.14')  | 3.14
bin()   | 轉換為二進位字串 | bin(99) | '0b1100011'
oct()   | 轉換為八進位字串 | oct(99) | '0o143'
hex()   | 轉換為十六進位字串 | hex(A9) | '0x63'
bool()  | 轉換為布林值    | bool(100) | True
str()   | 轉換為字串       | str(99) | '99'
chr()   | 轉換ASCII碼為字元 | chr(88) | 'X'
ord()   | 轉換字元為ASCII碼 | ord('X') | 88
<cr>
<center>**數值運算**</center>

名稱   |  功能  |  範例  |  結果
:----:|:-------|--------|-------|
round()  | 無條件捨去小數 | round(3.14) | 3
pow()    | 乘方計算      | pow(3,2)    | 9
abs()    | 絕對值        | abs(-99)  | 99
divmod() | 除法求商及餘數 | divmod(55,23) | (2,9)
max()    | 最大值        | max([1, 3, 5, 7]) | 7
min()    | 最小值        | min([3, 5, 2, 8]) | 2
sum()    | 總和          | sum([2, 5, 3, 4]) | 14

In [28]:
ord('X')

88

In [29]:
#round()內建函式
print(round(3.14159))
print(round(3.14159, 2))

3
3.14


In [30]:
#chr()與ord()內建函式
for i in range(65, 91):
    print(chr(i), end=" ")
print('\n')
for j in ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','s','t','r','u','v','x','y','z']:
    print(ord(j), end=" ")
print('\n')

A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 

97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 115 116 114 117 118 120 121 122 



## 6.3 自訂函式

使用者自訂函式是程式開發的建構元件，是構成程式結構的基本單元，針對程式執行的功能或特定的處理流程，可以寫成函式，函式提供多次使用，甚至可以被其他程式呼叫。程式以函式組成，使程式容易維護，容易理解，並可支援模組化設計。由於函式可以獨立開發，所以使程式開發可以分工，以加快開發時程。

### 定義函式

```python
def 函式名稱(輸入參數列):
    程式敘述1
    程式敘述2
    .....
    [return]
```

使用者自訂函式須使用`def`關鍵字，函式的名稱須有意義，名稱後加上小括號，括號內包含輸入參數列，可以沒有參數或多個參數。如有多個參數，須用逗點「,」隔開輸入參數。例如，`def func(x1, x2, x3)`。小括號後一定要加「:」符號。函式內部的程式敘述要遵守內縮規則。如果函式有輸出項，則須要`return`關鍵字。函式的執行碰到`return`關鍵字，則停止執行，跳出函式。

In [1]:
#範例一
def sayHello():
    print("Hello")
for i in range(5):
    sayHello()

Hello
Hello
Hello
Hello
Hello


In [4]:
#範例二：二整數相加(有輸入參數)
def addTwo(x, y):
    print("{}+{}={}".format(x, y, x+y))
addTwo(5, 6)
addTwo(100, 500)

5+6=11
100+500=600


In [8]:
#範例三：回傳二整數相加(有輸入參數)
def sqRoot(x):
    return x**(1/2)
n = int(input("輸入一個數字："))
print("{}的平方根為{}".format(n, sqRoot(n)))


輸入一個數字：11
11的平方根為3.3166247903554


### 區域及全域變數

在函式內部出現的變數（包含輸入參數）稱為區域變數(Local variable)。區域變數的有效範圍只有在函式內部，離開函式就無法作用。全域變數則是在函式外定義的變數，當函式內定義與全域變數相同名稱的區域變數時，若要存取全域變數時，必須在函式內先宣告`global 全域變數`，否則變數的存取以區域變數為主。

In [9]:
def f():
    print(s)
s = "I am handsome boy" #s是全域變數
f()

I am handsome boy


In [11]:
def f(): 
    s = "Me too." #s為區域變數
    print(s)

s = "I hate spam." #s為全域變數
f()
print(s) #s為全域變數

Me too.
I hate spam.


In [13]:
def f(): 
    print(s)
    s = "Me too." #s為區域變數
    print(s)

s = "I hate spam."  #s為全域變數
f()
print(s)  #s為全域變數

UnboundLocalError: local variable 's' referenced before assignment

In [19]:
def f():
    global s
    print(s) #s為全域變數
    s = "That's clear." #改變全域變數
    print(s) #s為全域變數


s = "Python is great!" #s為全域變數
f()
print(s) #s為全域變數

Python is great!
That's clear.
That's clear.


In [18]:
def f():
    a = "I am globally not known"
    print(a) 

f()
print(a)

I am globally not known


NameError: name 'a' is not defined

In [20]:
def foo(x, y):
    global a
    a = 42
    x,y = y,x
    b = 33
    b = 17
    c = 100
    print(a,b,x,y)

a,b,x,y = 1,15,3,4
foo(17,4)
print(a,b,x,y)

42 17 4 17
42 15 3 4


## 6.4 函式庫

### 函式庫呼叫

開發大型軟體時，程式的結構非常重要，可以將所開發的軟體檔案有效管理。python的程式檔案管理可以使用模組(modules)及套件(packages)來完成。實務上，模組包含軟體運作所需的函式(functions)、類別(class)、或變數定義(variables)。模組可視為python檔案，套件則可視為儲存管理模組的目錄，包含模組及次套件(次目錄)。

如果在某一個套件下，有二個模組，`m1.py`及`m2.py`。其中，`m1`模組中有`f1()`及`f2()`函式。在`m2`模組中要用到`f1()`函式，則需要使用`import`關鍵字由`m2`模組中導入`f1`函式。

```python
#在m2.py
from m2 import f1, f2

f1()
```

將m1及m2二個模組放在一個套件下, p1。則套件的結構如下

```
p1/
├ __init__.py 
├ m1.py
⎿ m2.py
```
`__init__.py`一定要有，空的內容也沒關係。


呼叫模組的方式有以下三種，

方式一：
>`import random` 

>`print(random.randint(1,100))`

方式二：
>`from random import randint`

>`print(randint(1,100))`

方式三：
> `import random as rr`

> `print(rr.randint(1, 100))`

當程式中導入一個模組，需要經由模組路徑來找到模組，直譯器首先搜尋內建模組，如果未尋獲，則搜尋sys.path變數中是否存有模組所在的資料夾。如果，導入模組的python程式(module_test.py)在/home/test1目錄路徑，而模組(m2.py)在/home/test2目錄路徑，內涵f1及f2函式，在module_test.py中可以做以下處理
```python
import sys
sys.path.append("/home/test2")
imort m2
f1()
```

`dir()`函式可以用來顯示有關模組的所有內容項目，如`dir(math)`可以顯示所有關於math模組的內容，其中`__name__`為模組名稱。

In [1]:
import math
print(dir(math))

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


In [23]:
import math
math.ceil(5.0)

5

### 數學函式庫

`math`模組提供許多有關數學的函式，使用數學模組前先導入`import math`

**數字理論與表示函式**

| 數學函式          |  意義           | 範例     | 結果  |
|-----------------|-------------------|-------|-------|
| math.ceil(x)      |  比x大的整數      | math.ceil(3.2)  |  4     |
| math.fabs(x)      |  絕對值      | math.fabs(-3)  |    3   |
| math.factorial(x) |  階層數   | math.factorial(5) |   120    |
| math.floor(x)     |  比x小的整數               | math.floor(3.2) |  3     |
| math.fmod(x,y)    |   求浮點數x/y的商及餘數              | math.fmod(5.3, 2.3) |  0.7     |
| math.gcd(a,b)    |   求最大公因數              | math.gcd(12, 44)      |   4    |
| math.trunc(x) |   無條件捨去小數              | math.trunc(3.2)      |    3   |

**指數及對數函式**

| 數學函式          |  意義           | 範例     | 結果  |
|-----------------|-------------------|-------|-------|
| math.exp(x)      |  指數函數   |  math.exp(1)   | 2.718281828459045 |
| math.log(x[, base])   |  對數函數  | math.log(10, 2) = math.log2(10) | 3.3219280948873626 |
| math.pow(x, y)   |  x的y次方   | math.pow(3,3)    | 27.0 |
| math.sqrt(x)   |  開根號函數  |  math.sqrt(4.4) | 2.0976176963403033  |

**三角函式**

| 數學函式          |  意義           | 
|-----------------|-------------------|
| math.acos(x)  |  反餘弦函數         |
| math.asin(x)    |  反正弦函數         |
| math.atan(x)    |  反正切函數         | 
| math.cos(x)     |  餘弦函數         | 
| math.sin(x)     |  正弦函數         | 
| math.tan(x)     |  正切函數         | 

**角度轉換**

| 數學函式          |  意義           | 範例     | 結果  |
|-----------------|-----------------|-------|-------|
| math.degrees(x) | 徑度轉換角度   |  math.degrees(math.pi/2) | 90.0 |
| math.radians(x) | 角度轉換徑度   |  math.degrees(90.0) |  1.5707963267948966 |

<center>$\frac{math.pi}{2} $ = 1.5707963267948966 </center>

**數學常數**

| 數學函式          |  意義           | 範例     | 結果  |
|-----------------|-------------------|-------|-------|
| math.pi      |  圓周率      | π = 3.141592… |     |
| math.e      |  自然數      | e = 2.718281…  |      |
| math.nan | not a number    | |      |
| math.inf | positive infinity | |


### 亂數函式庫

random 模組

|  亂數函數  |  函數意義  |
|----------|-----------|
| random.seed(x) | 設定亂數產生器種子 |
| randrange(start, stop[, step])	| 在範圍中產生一個隨機整數 |
| randint(a, b)	| 在a, b之間產生一個隨機整數 |
| choice(seq)	| 在一個非空的序列中隨機選取一個元素 |
| shuffle(seq) | 序列元素洗牌 |
| sample(population, k)	| 在一個母體序列(population sequence)中，隨機選取k個元素。 |
| random()	| 在0, 1範圍中產生一個隨機亂數 |
| uniform(a, b)	| 在a, b之間產生一個均質亂數 |
| normalvariate(mu, sigma)	| 常態分布中取一亂數 |

## 參考資料

1. [Python course by Bernd Klein](https://www.python-course.eu/python3_functions.php)
2. [Python for Everybody by Dr. Chuck](https://www.py4e.com/lessons/functions)
3. [Learn Python Programming at Programiz](https://www.programiz.com/python-programming/function)